fendermint/vm/topdown/src/sync.rs

```diff
@@ -242,106 +239,217 @@ async fn reset_cache<T: ParentFinalityStateQuery + Send + Sync + 'static>(
     Ok(())
 }
 
-/// Obtains the last recorded height from provider cache or from last committed finality height.
-async fn last_recorded_height(
+/// A util struct that tracks the last recorded height
+enum LastRecordedBlock {
+    FilledBlock {
+        height: BlockHeight,
+        hash: BlockHash,
+    },
+    NullBlock(BlockHeight),
+    Empty,
+}
+
+impl LastRecordedBlock {
+    fn filled(height: BlockHeight, hash: BlockHash) -> Self {
+        Self::FilledBlock { height, hash }
+    }
+
+    fn null(height: BlockHeight) -> Self {
+        Self::NullBlock(height)
+    }
+
+    fn empty() -> Self {
+        Self::Empty
+    }
+}
+
+/// Getting the last recorded block height/hash
+async fn last_recorded_data(
     provider: &Arc<Toggle<CachedFinalityProvider<IPCProviderProxy>>>,
 ) -> anyhow::Result<Option<(BlockHeight, BlockHash)>> {
-    let result = atomically(|| {
-        let h = if let Some(h) = provider.latest_height_hash()? {
-            Some(h)
-        } else if let Some(f) = provider.last_committed_finality()? {
-            Some((f.height, f.block_hash))
-        } else {
-            None
-        };
-        Ok(h)
-    })
-    .await;
+    match atomically(|| last_recorded_block(provider)).await {
+        LastRecordedBlock::Empty => Ok(None),
+        LastRecordedBlock::FilledBlock { height, hash } => Ok(Some((height, hash))),
+        LastRecordedBlock::NullBlock(height) => {
+            // Imaging the list of blocks as follows:
+            //
+            // F0  B0  B1  N0  N1  B2  B3  B4
+            //
+            // where F0 represents the last committed finality, B* represents non-null blocks and
+            // N* represents null blocks.
+            //
+            // Currently the last recorded block is N1, so the next block to sync in parent is B2.
+            // The response from getting block hash at height B2 from fvm eth apis would return:
+            //
+            // Block height: B2, Block hash: hash(B2), Parent block hash: hash(B1)
```

Comment:
Thanks you for including these examples!

I think I misunderstood what a null block was, I thought it would be something like this:
```
B0 B1 N2 N3 B4 B5
```
That is, at height 2 and 3 there wouldn't be any blocks at all, and without a reorg that would not change. I was curious what the API would return for `getBlockByHeight(2)` and `getBlockByHeight(4)` in this case. Are you saying that it will actually return `B4` in my example at height 2?

Reply:
If you `getBlockByHeight(2)`, it would panic and error is null round sth. But if you `getBlockByHeight(4)`, it returns the block hash at height 4, but the parent block hash is actually `B1`'s block hash.